### ISSUE TITLE
Mono.usingWhen operator not respecting onError behaviour 

### ISSUE THREAD

**Issue by hfgbarrigas** [reactor/reactor-core#1832]
Although async clean up handlers are correctly triggered, if they cause an error, the error is bubbled up but the onError behaviour is not being respected on the returning mono. For example, if onComplete handler returns a Mono.error, a doOnSuccess applied on the returning mono from Mono.usingWhen gets called.

## Expected Behavior

Successful operators should not be called if an error was thrown within async handlers.

## Actual Behavior

Successful operators are being called in spite of an error being thrown and intercepted by usingWhen.

## Steps to Reproduce

```java
@Test
    public void usingWhenTest() {
        Mono.usingWhen(Mono.just("data"),
                (Function<String, Mono<String>>) s -> Mono.just("resourceClosure"),
                s -> {
                    System.out.println("Complete resource");
                    return Mono.error(new RuntimeException("Complete Exception"));
                },
                s -> {
                    System.out.println("Error resource");
                    return Mono.empty();
                },
                s -> {
                    System.out.println("Cancel resource");
                    return Mono.empty();
                })
                .doOnSuccess(s -> System.out.println("Success: " + s))
                .doOnError(err -> System.out.println("Error: " + err))
                .map(s -> {
                    System.out.println("Map operator: " + s);
                    return s;
                })
                .subscribe();
    }
```

## Your Environment

* Reactor version(s) used: spring boot dependencies 2.2.0.M3

**Comment by hfgbarrigas** [comment]
I can confirm the same behaviour on Flux.

Example:

```
@Test
    public void usingWhenTest2() {
        Flux.usingWhen(Mono.just("data"),
                (Function<String, Flux<String>>) s -> Flux.just("resourceClosure", "resourceClosure2"),
                s -> {
                    System.out.println("Complete resource");
                    return Mono.error(new RuntimeException("Complete Exception"));
                },
                s -> {
                    System.out.println("Error resource");
                    return Mono.empty();
                },
                s -> {
                    System.out.println("Cancel resource");
                    return Mono.empty();
                })
                .doOnError(err -> System.out.println("Error: " + err))
                .map(s -> {
                    System.out.println("Map operator: " + s);
                    return s;
                })
                .subscribe();
    }
````

**Comment by simonbasle** [comment]
@hfgbarrigas seems that the behavior is not the same on `Flux` though... You don't have a `doOnComplete` here, but if you put up one, you'll see that only the `doOnError` is invoked.

For `Mono`, there is definitely an issue. That is coming from the fact that most if not all mono operators assume that receiving an `onNext` implies `onComplete`. This short-circuiting logic is implemented in the `doOnSuccess` operator, which invokes its `Runnable` as soon as it receives the `onNext("resourceClosure")`, hence the success.

One way the mono version could be modified to accommodate this would be to defer the emission of the `onNext` until the `completeAsyncHandler` terminates. If the async handler fails with an error, then the value would NOT be emitted, though it could be passed along the error to downstream (eg. `new CompleteAsyncHandlerException("msg", valueThatWasEmitted, exceptionFromTheAsyncHandler)`).

The other way would be to simply drop any error thrown inside the `completeAsyncHandler` when a value has already been emitted, but I think this is a more dangerous assumption (think "db commit that fails": do you prefer the onNext to be deferred until the confirmed end of the commit or for a commit failure to be silenced?).

cc @mp911de, @smaldini

**Comment by simonbasle** [comment]
> Same issue here. We should emit the value and then emit the error signal.

But that violates the `Mono` contract, doesn't it?

**Comment by mp911de** [comment]
I had always the understanding that Mono emits zero to one data signals and then either completion or error. Not really that success is tied to the data signal.

**Comment by hfgbarrigas** [comment]
@simonbasle indeed the onComplete is not called. I was under the impression that data signals were deferred until async handlers completed, which apparently is not the case. If that is by design, all is well with Flux :)

**Comment by simonbasle** [comment]
planned fix for this one: unlike for `Flux#usingWhen`, `Mono#usingWhen` will defer emitting the value until the `asyncComplete` itself completes, so that an async error translates downstream to **only** `onError` signal. The emitted value will be discarded in that case. Otherwise, the downstream will correctly receive `onNext+onComplete` signals. But never `onNext+onError`.

**Comment by simonbasle** [comment]
confirming from discussions with @smaldini that `Flux#usingWhen` should be left as is: even though the two versions would behave slightly differently, it makes sense that the `Mono` _specialization_ has a _specialized_ behavior... The API already takes a slightly different `resourceClosure` (Flux takes a `Publisher`, Mono takes a `Mono`).

we incidentally need to make sure that the javadoc explains the difference AND that it makes it very clear that the only part _really_ protected by the async handlers is the one inside the `resourceClosure` `Function`.

we need to also clarify in a very unambiguous manner that `onNext+onError` is not supported for `Mono` (eg. in the head javadoc of `Mono.java`).

**Comment by hugocosta88** [comment]
@simonbasle do you have any idea about the release date for this issue in 3.2.12.RELEASE?

**Comment by simonbasle** [comment]
@hugocosta88 3.2.12.RELEASE should be released next week

**Comment by simonbasle** [comment]
@hugocosta88 note that I kept this issue in 3.2.12 milestone by mistake. this has been fixed on `master`, so will be part of `3.3.0.RC1`, also next week

**Comment by hugocosta88** [comment]
Ok. Thanks for the update @simonbasle
